### ISSUE TITLE
Confused about boxes for localization

### ISSUE THREAD

**Issue by luckyuho** [pierluigiferrari/ssd_keras#25]
Hi pierluigiferrari, thank you very much. after I study "train_ssd7", there are some process I am confused, hope you can help me. In "keras_ssd7.py", do we use both "only boxes" and "anchor_box"( with "variance box") to learn localization? what I learn from your code is the process as described below. 
We use anchor_box to do SSDEncoder to match each layer from ground truth. But in "ssd_box_encode_decode_utils.py"->"def encode_y"->"y_encoded" which we use "y_encode_template" to save the match box for "y_encoder", however, (1)I do not find "y_encode_template" any change in "def encode_y". 
After getting y_true from generator_anchor_boxes rather than anchor box, we compute loss for  localization with only boxes without the others. (2)Does this mean what we actually train is "only boxes"? And I have no idea we use "anchor box" and "variance box" for what
(3)Besides, I really have no idea what the variance box function is. If there are all 1 value or [0.1 0.3 0.5 0.7], what does it really mean?
(4)And the last one I do not understand is why do we need 4+4+4 boxes to predict localization instead only 4 coordinate box to learn for that? Is there any augmentation for that?
I'm sorry about I am confused about your hard work on explanation.
But I learn a lot from your great work, thank you!

**Comment by pierluigiferrari** [comment]
I don't think I fully understand the questions, but here are a few points regarding what I believe is at least part of what you're asking:

1. The `AnchorBoxes` layer does not participate at all in the training of the model. Its only purpose is to just output the anchor box coordinates and variances so that `decode_y()` or `decode_y2()` can decode the raw model prediction tensor without needing any other information. This is also why the model output tensor's last axis has length `n_classes + 4 + 4 + 4`. The last 8 elements of the last axis are just the four anchor box coordinates and the four variances for each box and as mentioned before, they are not relevant to the training, but only to decode the model output at inference time. I also recommend reading the documentation and inline comments of `AnchorBoxes`, which should help understand why the model output tensor has this particular shape.
2. The variances are just scaling factors for the ground truth box coordinates. They allow you to scale the individual coordinate offsets independently. I recommend you take a look at the comments at the bottom of the definition of `encode_y()` to better understand what they do. The coordinate offsets are being divided by the variances. For example, the variances chosen in the original SSD300 are `[0.1, 0.1, 0.2, 0.2]`, meaning that the `(cx, cy)` values are up-scaled by a factor of 10 and the `(w, h)` values are upscaled by a factor of 5. Among other things this means that the box center coordinates are weighted stronger than the width/height values. The idea behind trying different values for these variances is simply to see if the model learns better with a certain set of values.

**Comment by luckyuho** [comment]
oh my god,
thank you a lot, pierluigiferrari!!!
I think I finally understand your comments!!
It was my fault, I totally misunderstood it.
I totally suddenly see the light.
Thanks for your help!!!

**Comment by pierluigiferrari** [comment]
The idea is this:

We have four scalar anchor box coordinates, cx, cy, w, h. For each of these four coordinates, the desired prediction for a given ground truth box could be either larger or smaller than the respective anchor box coordinate. For each of these four coordinates, we want the model to predict positive offsets in one direction and negative offsets in the other direction. And we want the predicted offsets to be relative to the respective absolute coordinate values of the anchor box. The chosen formula for the width and height fulfills both of these criteria:

`ln(g/d) = ln(g) - ln(d) > 0`  if `g > d`, `< 0` if `g < d`  and  `ln(a*g/a*d) = ln(g/d)`  for any real number `a`

Whether or not this is the best transformation of the target coordinates for the model to learn optimally is a different story, but it is at least one possible transformation that works well.

**Comment by luckyuho** [comment]
Oh oh oh,
I think I have learned a lot and known something.
Really very thanks you!!!
